### PR TITLE
Do not classify Infinity and NaN

### LINKED ISSUE #42022: JS/TS semantic highlighting for keywords (readonly global vars)
*TS Template added by @mjbvz*

**TypeScript Version**: 4.1.2

**Search Terms**
- semantic highlighting
- constant
- readonly

---

vscode 1.52.0

Code:

```js
undefined;
Infinity;
NaN;
```

Result:

Semantic highlighting recolors `Infinity` and `NaN` from `constant.language` to `variable`.

Expected:

All three identifiers should be colored identically.

### PR DESCRIPTION
**Pull request by jeanp413**
Fixes #42022

### CODE REVIEW

File: src/services/classifier2020.ts [hunk 1]
```diff
@@ -225,6 +225,10 @@ namespace ts.classifier.v2020 {
         return (isQualifiedName(node.parent) && node.parent.right === node) || (isPropertyAccessExpression(node.parent) && node.parent.name === node);
     }
 
+    function isInfinityOrNaNString(name: __String): boolean {
+        return name === "Infinity" || name === "-Infinity" || name === "NaN";
+    }
+
```
Comment (jeanp413) [inline src/services/classifier2020.ts:231]: Not sure if there's a better way to check for `Infinity` and `NaN`, I copied this from `checker.ts` line 26202.
Let me know if there's a better way to fix this.

File: src/services/classifier2020.ts [hunk 2]
```diff
@@ -225,6 +225,10 @@ namespace ts.classifier.v2020 {
         return (isQualifiedName(node.parent) && node.parent.right === node) || (isPropertyAccessExpression(node.parent) && node.parent.name === node);
     }
 
+    function isInfinityOrNaNString(name: __String): boolean {
+        return name === "Infinity" || name === "-Infinity" || name === "NaN";
```
Comment (DanielRosenwasser) [inline src/services/classifier2020.ts:229]: I think you need a more thorough check, but some quick feedback: there's no identifier named `-Infinity`.

Comment (jeanp413) [reply src/services/classifier2020.ts:229]: > I think you need a more thorough check

Do you have something in particular in mind? Not sure what else could be done as `Infinity` and `NaN` are parsed as `SyntaxKind.Identifier` and I also found this check used in `checker.ts`

Comment (orta) [reply src/services/classifier2020.ts]: I've switched this PR to use the typechecker's function - which solves this 👍🏻

**Review by orta — APPROVED** [review]

File: tests/cases/fourslash/semanticModernClassificationInfinityAndNaN.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+/// <reference path="fourslash.ts"/>
+
+//// Infinity;
+//// NaN;
+
+verify.semanticClassificationsAre("2020");
```
Comment (DanielRosenwasser) [inline tests/cases/fourslash/semanticModernClassificationInfinityAndNaN.ts:6]: Shouldn't this get a baseline generated?

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]
Sorry, I thought we'd have to expose something from the checker, but the logic is exactly the same as that of the type-checker. I think I'm going to need to back-pedal a bit and suggest to just check against `Infinity` and `NaN`, since exposing another API now seems overkill.

On that note, could you add these cases to the existing test?

```ts
// Regular properties

const obj1 = {
    Infinity: 100,
    NaN: 200,
    "-Infinity": 300
};

obj1.Infinity;
obj1.NaN;
obj1["-Infinity"];

// Shorthand properties

const obj2 = {
    Infinity,
    NaN,
}

obj2.Infinity;
obj2.NaN;
```

File: tests/cases/fourslash/semanticModernClassificationInfinityAndNaN.ts (hunk 3 above)
Comment (orta) [reply tests/cases/fourslash/semanticModernClassificationInfinityAndNaN.ts:6]: I thought this (and validated it) but the point is that it is ignored, so the lack of any classifications is actually a correct check

**Comment by orta** [comment]
Cool, improved the tests and made it only check for "Infinity" and not `-Infinity` (which didn't make a difference on the tests, so wasn't doing work anyway)

**Review by DanielRosenwasser — APPROVED** [review]

**Pull request merged**